File: csharp/StrangeCharactersRefactoring/Characters/CharacterFactory.cs
namespace Characters;

/**
 * Factory for commonly used lists of Characters
 */
public class CharacterFactory
{
    public static List<Character> StrangeCharacters()
    {
        var joyce = new Character("Joyce", "Byers");
        var jim = new Character("Jim", "Hopper");
        var mike = new Character("Mike", "Wheeler");
        var eleven = new Character("Eleven");
        var dustin = new Character("Dustin", "Henderson");
        var lucas = new Character("Lucas", "Sinclair");
        var nancy = new Character("Nancy", "Wheeler");
        var jonathan = new Character("Jonathan", "Byers");
        var will = new Character("Will", "Byers");
        var karen = new Character("Karen", "Wheeler");
        var holly = new Character("Holly", "Wheeler");
        var steve = new Character("Steve", "Harrington");
        var mindflayer = new Character("Mindflayer", isMonster:true);
        var demagorgon = new Character("Demagorgon", isMonster:true);
        var demadog = new Character("Demadog", isMonster:true);
        var mrwhatsit = new Character("MrWhatsit", isMonster:true);
        
        joyce.AddChild(jonathan);
        joyce.AddChild(will);
        jim.AddChild(eleven);
        karen.AddChild(nancy);
        karen.AddChild(mike);
        karen.AddChild(holly);
        
        eleven.SetNemesis(demagorgon);
        will.SetNemesis(mindflayer);
        dustin.SetNemesis(demadog);
        holly.SetNemesis(mrwhatsit);

        return new List<Character>()
        {
            joyce,
            jim,
            mike,
            will,
            eleven,
            dustin,
            lucas,
            nancy,
            jonathan,
            mindflayer,
            demagorgon,
            demadog,
            karen,
            holly,
            steve,
            mrwhatsit
        };
    }

    public static List<Character> WheelerFamily()
    {
        var karen = new Character("Karen", "Wheeler");
        var mike = new Character("Mike", "Wheeler");
        var nancy = new Character("Nancy", "Wheeler");
        var holly = new Character("Holly", "Wheeler");

        karen.AddChild(nancy);
        karen.AddChild(mike);
        karen.AddChild(holly);

        return new List<Character>()
        {
            karen,
            mike,
            nancy,
        };
    }
}

File: java/src/main/java/codingdojo/CharacterFactory.java
package codingdojo;

import java.util.Arrays;
import java.util.Collection;

public class CharacterFactory {
    public static Collection<Character> allTestData() {
        var joyce = new Character("Joyce", "Byers");
        var jim = new Character("Jim", "Hopper");
        var mike = new Character("Mike", "Wheeler");
        var eleven = new Character("Eleven");
        var dustin = new Character("Dustin", "Henderson");
        var lucas = new Character("Lucas", "Sinclair");
        var nancy = new Character("Nancy", "Wheeler");
        var jonathan = new Character("Jonathan", "Byers");
        var will = new Character("Will", "Byers");
        var karen = new Character("Karen", "Wheeler");
        var holly = new Character("Holly", "Wheeler");
        var steve = new Character("Steve", "Harrington");
        var mindflayer = new Character("Mindflayer", true);
        var demagorgon = new Character("Demagorgon", true);
        var demadog = new Character("Demadog", true);
        var mrwhatsit = new Character("MrWhatsit", true);

        joyce.addChild(jonathan);
        joyce.addChild(will);
        jim.addChild(eleven);
        karen.addChild(nancy);
        karen.addChild(mike);
        karen.addChild(holly);

        eleven.setNemesis(demagorgon);
        will.setNemesis(mindflayer);
        dustin.setNemesis(demadog);
        holly.setNemesis(mrwhatsit);

        return Arrays.asList(
                joyce,
                jim,
                mike,
                will,
                eleven,
                dustin,
                lucas,
                nancy,
                jonathan,
                mindflayer,
                demagorgon,
                demadog,
                karen,
                holly,
                mrwhatsit,
                steve);
    }

    public static Collection<Character> wheelerFamily() {
        Character karen = new Character("Karen", "Wheeler");
        Character mike = new Character("Mike", "Wheeler");
        Character nancy = new Character("Nancy", "Wheeler");
        Character holly = new Character("Holly", "Wheeler");

        karen.addChild(nancy);
        karen.addChild(mike);
        karen.addChild(holly);

        return Arrays.asList(karen, mike, nancy);
    }
}
